README.md

```diff
@@ -2,9 +2,147 @@
 
 >  A Hardware Abstraction Layer (HAL) for embedded systems
 
-## [Documentation]
+## [API reference]
 
-[Documentation]: https://japaric.github.io/embedded-hal/embedded_hal/index.html
+[API reference]: https://docs.rs/embedded-hal
+
+## How-to: add a new trait
+
+This is the suggested approach to adding a new trait to `embedded-hal`
+
+### Discussion
+
+Ideally, before proposing a new trait, or set of traits, you should create an issue where the use
+cases and requirements of the trait(s) are discussed.
+
+These issues will be labeled as `discussion`s in the issue tracker.
+
+### Proposing a trait
+
+Once there's consensus on the requirements of the trait(s) a new issue, or a PR, with a proposal
+should be opened. The proposal should include the actual trait definitions as well as a link to the
+issue with previous discussion, if there was one.
+
+If the proposal includes more than one alternative then there should be further discussion to try to
+single out the best alternative.
+
+These issues / PRs will be labeled as `proposal`s in the issue tracker.
+
+### Testing period
+
+If there are no objections to the proposal the new trait(s) will land behind the "unproven" Cargo
+feature and an issue about the new trait(s) will be created. If the proposal includes several
+alternatives and a single one couldn't be chosen as the best then each alternative will land behind
+a different Cargo feature, e.g. "alt1" or "alt2".
+
+The traits will undergo a testing period before they move into the set of proven traits. During
+this period users are encouraged to try to implement the unproven traits for their platforms and to
+build drivers on top of them. Problems implementing the trait(s) as well as successful
+implementations should be reported on the corresponding issue.
+
+To leave the unproven state at least *two* implementations of the trait(s) for different platforms
+(ideally, the two platforms should be from different vendors) and *one* generic driver built on top
+of the trait(s) must be demonstrated. If, instead, reports indicate that the proposed trait(s) can't
```

Comment:
There might be cases where a trait is purely internal and/or may not have any drivers using it, like timers, ADCs, RNGs... in that case a full example demonstrating the use should suffice, I think.

Reply:
I think that's a fair observation. Maybe the requirements should be worded less strongly and be treated as more of a guideline.